[OpenMP][libomptarget] Enable parallel copies via multiple SDMA engines

jplehr opened the pull request:
This enables the AMDGPU plugin to use a new ROCm 5.7 interface to dispatch asynchronous data transfers across SDMA engines.

The default functionality stays unchanged, meaning that all data transfers are enqueued into a H2D queue or an D2H queue, depending on transfer direction, via the HSA interface used previously.

The new interface can be enabled via the environment variable `LIBOMPTARGET_AMDGPU_USE_MULTIPLE_SDMA_ENGINES=true` when libomptarget is built against a recent ROCm version (5.7 and later).
As of now, requests are distributed in a round-robin fashion across available SDMA engines.

jhuber6 commented on openmp/libomptarget/plugins-nextgen/amdgpu/src/rtl.cpp:
```diff
@@ -130,6 +130,40 @@ Error iterateAgentMemoryPools(hsa_agent_t Agent, CallbackTy Cb) {
+                   hsa_agent_t SrcAgent, size_t Size, uint32_t NumDepSignals,
+                   const hsa_signal_t *DepSignals,
+                   hsa_signal_t CompletionSignal) {
+  static BoolEnvar OMPX_UseMultipleSdmaEngines{
```
This will emit a global constructor which we really don't want to have scattered around in the plugin. We need to initialize this in the AMDGPUDeviceTy or somewhere else.

jplehr replied:
Thanks for pointing that out. I moved the Envar into the AMDGPUDeviceTy and have added a boolean to the StreamTy, so we do not need to access the device for the value.

arsenm commented on openmp/libomptarget/plugins-nextgen/amdgpu/dynamic_hsa/hsa_ext_amd.h:
```diff
@@ -151,12 +151,10 @@ typedef struct hsa_amd_pointer_info_s {
   size_t sizeInBytes;
 } hsa_amd_pointer_info_t;
 
-hsa_status_t hsa_amd_pointer_info(const void* ptr,
-                                          hsa_amd_pointer_info_t* info,
-                                          void* (*alloc)(size_t),
-                                          uint32_t* num_agents_accessible,
-                                          hsa_agent_t** accessible);
-
```
Can separately submit this NFC reformat

jplehr committed:
[OpenMP][libomptarget] Enable parallel copies via multiple SDMA engines

This enables the AMDGPU plugin to use a new ROCm 5.7 interface to
dispatch asynchronous data transfers across SDMA engines.

The default functionality stays unchanged, meaning that all data
transfers are enqueued into a H2D queue or an D2H queue, depending on
transfer direction, via the HSA interface used previously.

The new interface can be enabled via the environment variable
LIBOMPTARGET_AMDGPU_USE_MULTIPLE_SDMA_ENGINES=true when libomptarget is
built against a recent ROCm version (5.7 and later).
As of now, requests are distributed in a round-robin fashion across
available SDMA engines.

jhuber6 commented on openmp/libomptarget/plugins-nextgen/amdgpu/src/rtl.cpp:
```diff
@@ -2623,6 +2652,8 @@ struct AMDGPUDeviceTy : public GenericDeviceTy, AMDGenericDeviceTy {
         });
   }
 
+  bool getUseMultipleSdmaEngines() { return OMPX_UseMultipleSdmaEngines; }
```
Having this be a `getter` is weird, maybe just `useMultipleSdmaEngines`.

jplehr replied:
That's a good point. I also added `const` here.

jhuber6 commented on openmp/libomptarget/plugins-nextgen/amdgpu/src/rtl.cpp:
```diff
@@ -130,6 +130,38 @@ Error iterateAgentMemoryPools(hsa_agent_t Agent, CallbackTy Cb) {
+  static int SdmaEngine = 1;
+
+  // This call is only avail in ROCm >= 5.7
+  hsa_status_t S = hsa_amd_memory_async_copy_on_engine(
```
What happens if someone used dynamically opened HSA to an older version? Do we crash?

jplehr replied:
I think if users are going to set the env var to `true` than we would crash. I can do a build against an older ROCm version to see what happens.
Remember, this is not default behavior, but has to be enabled explicitly.

Would it be acceptable to note that in the Documentation (that I still have to add)? Since we have limited capability to check for which HSA version we are running, this is indeed a bit of a pain point.

jhuber6 replied:
Alright, since it's opt-in it's safe.

jhuber6 commented on openmp/libomptarget/plugins-nextgen/amdgpu/src/rtl.cpp:
```diff
@@ -130,6 +130,38 @@ Error iterateAgentMemoryPools(hsa_agent_t Agent, CallbackTy Cb) {
                        "Error in hsa_amd_agent_iterate_memory_pools: %s");
 }
 
+/// Dispatches an asynchronous memory copy
```
```suggestion
/// Dispatches an asynchronous memory copy.
```

jhuber6 commented on openmp/libomptarget/plugins-nextgen/amdgpu/src/rtl.cpp:
```diff
@@ -130,6 +130,38 @@ Error iterateAgentMemoryPools(hsa_agent_t Agent, CallbackTy Cb) {
+    hsa_status_t S =
+        hsa_amd_memory_async_copy(Dst, DstAgent, Src, SrcAgent, Size,
+                                  NumDepSignals, DepSignals, CompletionSignal);
+    return Plugin::check(S, "Error in hsa_amd_memory_async_copy");
```
Possible for us to print the error? IIRC there's some automatic handling here if there's `%s` or something in the error message? I forget exactly.

jhuber6 commented on openmp/libomptarget/plugins-nextgen/amdgpu/src/rtl.cpp:
```diff
@@ -130,6 +130,38 @@ Error iterateAgentMemoryPools(hsa_agent_t Agent, CallbackTy Cb) {
+      HSA_AMD_INTERFACE_VERSION_MINOR >= 2)
+  return Plugin::error("Async copy on selected SDMA requires ROCm 5.7");
+#else
+  static int SdmaEngine = 1;
```
This is a shared variable. What happens when multiple threads potentially run through here?

jplehr replied:
That's a good point. In most instances, this function is called from a region that is protected via a `lock_guard`, so IIRC for those cases this is protected and there should not be two threads here at the same time.
But I have to look more into one case, where this may race.

jhuber6 replied:
We could probably use an atomic RMW just to be safe in any case.

jplehr replied:
So, I looked into it and we do have one code path that is not under a lock. I made this an atomic and do RMW.

With this solution we could still run into a scenario where two threads read the same value and dispatch to the same SDMA engine. While not desirable, it's not a correctness issue, and I think the probability is quite low.

jplehr committed:
fixup! [OpenMP][libomptarget] Enable parallel copies via multiple SDMA engines

jplehr committed:
fixup! fixup! [OpenMP][libomptarget] Enable parallel copies via multiple SDMA engines

jhuber6 reviewed (approved):
LG, thanks for sticking with it. Shame there's no real way to check for the HSA version.

jplehr commented on openmp/libomptarget/plugins-nextgen/amdgpu/src/rtl.cpp:
```diff
@@ -130,6 +130,45 @@ Error iterateAgentMemoryPools(hsa_agent_t Agent, CallbackTy Cb) {
+                   const void *Src, hsa_agent_t SrcAgent, size_t Size,
+                   uint32_t NumDepSignals, const hsa_signal_t *DepSignals,
+                   hsa_signal_t CompletionSignal) {
+  if (UseMultipleSdmaEngines) {
```
This needs to be negated to maintain the current functionality per default and not rely on ROCm 5.7 functionality.
